### ISSUE TITLE
Using scalafmt from the editor can make Metals hang

### ISSUE THREAD

**Issue by kubukoz** [scalameta/metals#2408]
**Describe the bug**

I'm seeing ~600% CPU utilization from Metals after I format a large file a couple times. The exact amount is yet to be determined, but it's still the case even after disabling "format on save".

**To Reproduce**

- todo
- 

**Expected behavior**

The editor doesn't hang

**Screenshots**

![image](https://user-images.githubusercontent.com/894884/104972562-6ec08300-59f2-11eb-845f-fd663cdd8320.png)

![image](https://user-images.githubusercontent.com/894884/104972568-754efa80-59f2-11eb-8187-a6944148bbdb.png)

![image](https://user-images.githubusercontent.com/894884/104972583-7ed86280-59f2-11eb-9c7f-1c57b0ce3524.png)

**Installation:**

- Operating system: macOS
- Editor: Visual Studio Code
- Metals version: v0.9.8

**Additional context**

My metals server flags: `-Xmx1G`

**Search terms**

scalafmt, hangs, cpu, formatting, formatter

**Comment by ckipp01** [comment]
Thanks for the report @kubukoz! So I'm playing around with this a bit and I have a file (in Metals), which is 2500 lines long. I spammed format 30 times in a row, and then just waited.

<img width="1015" alt="Screenshot 2021-01-19 at 09 21 48" src="https://user-images.githubusercontent.com/13974112/105007075-d0571080-5a37-11eb-81c9-9d692430d9af.png">

However, a lot of the code in this file probably isn't that hard to parse compared some other code bases, so I'm curious in this file what you make heavy use of? Any specific language feature or library? Or if you've ever hit on this in any typelevel project or anything that have a huge file, feel free to point me in that direction. If not I'll start digging around for something.

Either way, there's no reason to see the numbers you're seeing. We need to get to the bottom of what's happening here.

**Comment by gabro** [comment]
It would be interesting to see whether the same issue is reproducible outside of the editor (i.e. with the Scalafmt CLI). IIRC there were performance issues in heavily nested code

**Comment by hedefalk** [comment]
I think I'm having the same problem. Seeing about 500-600% CPU

![image](https://user-images.githubusercontent.com/116520/106586021-44b9a580-6548-11eb-9f72-e62969c7ccde.png)


and since I have format-on-save turned on, the file save action seems to be intercepted and not saved to disk so sbt can't see the change which is very disruptive for me.

This notif at least lets me cancel the format and the save is done:

![image](https://user-images.githubusercontent.com/116520/106585999-3f5c5b00-6548-11eb-9263-3826c52842c6.png)

but after this Metals seems pretty much dead and 600% just continues on until I force kill it.

Question on that - how can I restart metals without reloading vscode? Seems I have no control over life-cycle there? Every metals command seems to fail after I've killed the metals java process?

**Comment by kubukoz** [comment]
@hedefalk you can kill the java process running metals, but I think then "Metals: restart server" won't work. Killing the process + "Developer: reload window" is probably your best bet.

BTW. Which JVM are you using to run metals?

Update from me: I have only seen this on one machine for now, on multiple projects. This mostly happened when formatting (which I stopped doing automatically in one of the most problematic projects, only formatting on demand) but not only then. I'm keeping my eyes out for future issues.

**Comment by tgodzik** [comment]
@hedefalk If it's a small file would you be able to post a reproduction? We don't need any specifics, only overall syntax, because that is what scalafmt operates on. Does it happen again if you try to format it? Maybe Metals is using too much memory?

There are two possible explanations for this:
- there is an issue in scalameta/scalafmt that causes formatter to never finish
- memory gets maxed and nothing can really work, because GC is running all the time

**Comment by kubukoz** [comment]
I posted some memory graphs from Metals on the gitter some time ago, it showed pretty consistently growing memory - suggesting a memory leak somewhere. Then scenario no.2 happens - this would be my guess...

**Comment by hedefalk** [comment]
@kubukoz Seems I'm running openjdk-11.0.1. I guess that's what locate-java-home sees, I should probably override this. Always use a shell.nix for cli usage.

@tgodzik I'll try to see if I can create a repro. This was just a 20-30 line file, but not standalone in anyway, pretty big sbt project so need to extract and see if I can make it happen again…

Could it be just as @kubukoz suggests, that it is just a growing memory leak? It seems it happens after a while of running. Currently I've been restructuring a build so I've been doing a lot of `sbt -Dbloop.export-jar-classifiers=sources bloopInstall` and "metals: connect to build server" as suggested elsewhere :)  Maybe that causes memory leaks too…?

**Comment by tgodzik** [comment]
> @kubukoz Seems I'm running openjdk-11.0.1. I guess that's what locate-java-home sees, I should probably override this. Always use a shell.nix for cli usage.
> 

That one has some additional issue I've seen, so for sure I would change it. 

> @tgodzik I'll try to see if I can create a repro. This was just a 20-30 line file, but not standalone in anyway, pretty big sbt project so need to extract and see if I can make it happen again…

If it's too much effort then don't worry about it. It's more probably the memory leak.

> Could it be just as @kubukoz suggests, that it is just a growing memory leak? It seems it happens after a while of running. Currently I've been restructuring a build so I've been doing a lot of `sbt -Dbloop.export-jar-classifiers=sources bloopInstall` and "metals: connect to build server" as suggested elsewhere :) Maybe that causes memory leaks too…?

That should not cause it as we are just running a separate process. I need to delve into that at some point, but haven't been able to get proper heap dump.

**Comment by pvid** [comment]
Hi! I am also running into this issues from time to time. I am also using the format on save feature. I remembered to capture a heap dump before restarting the process today. I am attaching a screenshot from Eclipse MAT. It seems pretty conclusive - several hundrens of `scala.tools.nsc.interactive.PresentationCompilerThread` coming from two different classloaders


![Screenshot 2021-02-03 at 17 55 40](https://user-images.githubusercontent.com/32688476/106781823-c513ff80-6649-11eb-9df0-0ec0535b241d.png)

each of them is from its own `MetalsGlobal` instance 

![Screenshot 2021-02-03 at 18 07 46](https://user-images.githubusercontent.com/32688476/106782689-c265da00-664a-11eb-91ad-1e50a920ee69.png)

but there are just a few `ScalaCompilerAccess` instances

![Screenshot 2021-02-03 at 18 20 43](https://user-images.githubusercontent.com/32688476/106784461-a4997480-664c-11eb-92af-bc0c72769acd.png)

I suspect that there is an error in the way that compiler shutdown is handled in [CompilerAccess](https://github.com/scalameta/metals/blob/main/mtags/src/main/scala/scala/meta/internal/pc/CompilerAccess.scala).

**Comment by kubukoz** [comment]
@pvid do you have your `metals/metals.log`? Are you seeing a lot of this, by chance?

> INFO: compiler crashed due to an error in the Scala compiler, retrying with new compiler instance.

**Comment by tgodzik** [comment]
@pvid Looks like this might be the problem, although I don't see exactly why this is happening :thinking:  

Especially weird is the name of the thread, we actually replace it with our own implementation and it should be named something like:
`"Scala Presentation Compiler [Metals]"`

**Comment by tgodzik** [comment]
It looks like the thread that we replace never really dies :O

**Comment by pvid** [comment]
> @pvid do you have your `metals/metals.log`? Are you seeing a lot of this, by chance?
> 
> > INFO: compiler crashed due to an error in the Scala compiler, retrying with new compiler instance.
> 
> This capture seems really good to have btw :D

yeah, I see some of those

> @pvid Looks like this might be the problem, although I don't see exactly why this is happening 🤔
> 
> Especially weird is the name of the thread, we actually replace it with our own implementation and it should be named something like:
> `"Scala Presentation Compiler [Metals]"`

I think the theads with `"Scala Presentation Compiler [Metals]"` are instances of [MetalsGlobalThread](https://github.com/scalameta/metals/blob/main/mtags/src/main/scala-2/scala/tools/nsc/interactive/MetalsGlobalThread.scala#L13). When connecting to a running Metals process using VisualVM there is one `"Scala Presentation Compiler [Metals]"` thread and a lot of `"Scala Presentation Compiler []"` that I believe come from [MetalsGlobal](https://github.com/scalameta/metals/blob/main/mtags/src/main/scala-2/scala/meta/internal/pc/MetalsGlobal.scala) which directly extends `scala.tools.nsc.interactive.Global`. `Global` has a third argument `projectName` which defaults to empty string

<img width="351" alt="Screenshot 2021-02-03 at 19 06 02" src="https://user-images.githubusercontent.com/32688476/106790130-694e7400-6653-11eb-82cb-e9f926a6a2cd.png">

**Comment by tgodzik** [comment]
@pvid it looks like on each restart we replace the compiler thread with Metals one, but it never really stops, just runs in the background.

https://github.com/scalameta/metals/blob/6a98caf6e07c8ac9f7e49f7b68bc3deac9d92468/mtags/src/main/scala-2/scala/tools/nsc/interactive/GlobalProxy.scala#L41

**Comment by tgodzik** [comment]
Seems that even when I use `Thread.stop()` the default presentation compiler Thread is still running. No idea how is that possible.

**Comment by tgodzik** [comment]
Let's see if https://github.com/scalameta/metals/pull/2476 will help, the compiler threads do nothing really now. But not 100% sure this is the main issue.

**Comment by tgodzik** [comment]
Closed via #2476

**Comment by kubukoz** [comment]
We'll see 🤣

**Comment by tgodzik** [comment]
I am staying optimistic :D You should be able to test with the newest snapshot @kubukoz

**Comment by kubukoz** [comment]
Looking back at this, IT HELPED. A LOT.

Metals is such a pleasure to use now.
